=title Python Flask: catch and handle exceptions in routes using errorhandler
=timestamp 2020-06-26T10:30:01
=indexes Flask, Exception, errorhandler, ZeroDivisionError
=status show
=books flask
=author szabgab
=archive 1
=comments_disqus_enable 0
=show_related 1

=abstract start

No matter how much your test your code, there might be an occassional exception raised during the execution of
one of the routes in your code. You cannot wrap all the code in a huge <b>try</b>-<b>except</b> block, but you
can use the built-in error-handling of Flask using the <b>errrohandler</b> hook that we have already
seen when we wanted to have a unified look for <a href="/flask-return-404">404 page not found</a>.

=abstract end

<h2>Catching the Exception</h2>

<code>
@app.errorhandler(Exception)
def server_error(err):
    app.logger.exception(err)
    return "exception", 500
</code>

Alternativelly, using a template:

<code>
@app.errorhandler(Exception)
def server_error(err):
    return render_template('crash.html'), 500
</code>


<h2>Sample application handling the Exception</h2>

<include file="examples/flask/catch_exception/app.py">

Run as:

<code>
FLASK_APP=app FLASK_DEBUG=1 flask run
</code>

Output on the console for regular pages.

<code>
[2020-06-26 10:33:22,903] INFO in app: main route
127.0.0.1 - - [26/Jun/2020 10:33:22] "GET / HTTP/1.1" 200 -
</code>


<h2>Oputput of the crash</h2>

Visiting the http://localhost:5000/crash URL shows a stack trace like this on the command line:

<code>
[2020-06-26 10:33:29,633] INFO in app: crash route
[2020-06-26 10:33:29,633] ERROR in app: division by zero
Traceback (most recent call last):
  File ".../flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File ".../flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File ".../flask/catch_exception/app.py", line 20, in crash
    b = 3 / a
ZeroDivisionError: division by zero
127.0.0.1 - - [26/Jun/2020 10:33:29] "GET /crash HTTP/1.1" 200 -
</code>

The user only sees the word "exception".

Of course insted of returning that word, we could have used <b>render_template</b> to
render any page. The status code returned by the page will be 500 as indicated by the
second value returned.


<h2>Catching specific exceptions</h2>

In the above examples we use the catch-all <b>Exception</b> class, but you can be more specific
about the exceptions and you can return different pages based on the error type.
You could do this by checking the <b>type</b> of the <b>err</b> variable, but a probably cleaner way
is to set up separate exception handler hooks for each exception type you'd like to deal with:

<include file="examples/flask/catch_specific_exception/app.py">

<h2>Logging the Exception</h2>

In the above examples we keept using the <hl>app.logger.exception(err)</hl> expression to log
the exception and the stack trace. If you are not interested in the stack-trace you can use
the <b>error</b> logging method: <hl>app.logger.error(err)</hl>.
